src/targets/gpu/include/migraphx/gpu/context.hpp

```diff
@@ -84,32 +86,93 @@ struct hip_device
         shared<rocblas_handle_ptr> rbhandle = nullptr;
     };
 
-    void add_stream() { streams.emplace_back(device_id); }
+    static hip_event_ptr create_event()
+    {
+        hipEvent_t event;
+        auto status = hipEventCreateWithFlags(&event, hipEventDisableTiming);
+        if(status != hipSuccess)
+            MIGRAPHX_THROW("Failed to creat event");
+        return hip_event_ptr{event};
+    }
+
+    void add_streams()
+    {
+        int num_of_streams = 1;
+        assert(streams.empty());
+        if(enabled(MIGRAPHX_DISABLE_NULL_STREAM{}))
+            num_of_streams = stream_info().num_of_streams();
+        for(int i = 0; i < num_of_streams; ++i)
+            streams.emplace_back(device_id);
+    }
 
     stream& get_stream() { return streams.at(current_stream); }
 
     void set_stream(std::size_t n) { current_stream = n; }
+    void create_events(int num_of_events)
+    {
+        for(int i = events.size(); i < num_of_events; ++i)
+            events.emplace_back(create_event());
+    }
+    void record_event(int event)
+    {
+        create_events(event + 1);
+        hipEventRecord(events.at(event).get(), streams.at(current_stream).get());
+    }
+
+    void wait_event(int event)
+    {
+        hipStreamWaitEvent(streams.at(current_stream).get(), events.at(event).get(), 0);
+    }
+
+    void stream_sync()
+    {
+        if(enabled(MIGRAPHX_DISABLE_NULL_STREAM{}))
+        {
+            int num_of_streams = streams.size();
+            if(num_of_streams > 0)
+            {
+                for(int i = 0; i < num_of_streams; i++)
+                    hipStreamSynchronize(streams.at(i).get());
+            }
+        }
+    }
 
     private:
     std::size_t device_id      = 0;
     std::size_t current_stream = 0;
     std::vector<stream> streams;
+    std::vector<shared<hip_event_ptr>> events;
 };
 
 struct context
 {
     context(std::size_t n = 0) : current_device(std::make_shared<hip_device>(n)) {}
 
-    hip_device& get_current_device()
+    hip_device& get_current_device() const
     {
         assert(current_device != nullptr);
         return *current_device;
     }
 
     hip_device::stream& get_stream() { return get_current_device().get_stream(); }
+    void set_stream(int n) const
```

Comment:
This should be non-const. Changing `get_current_device` with the overloads(explaned [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/175/files#diff-686d651b4eaae25eb72d9538d9e0ff38R151)) will let you make it non-const without tidy errors.

Reply:
Overloading get_current_device in return types does not compile.   Please merge, then you can change it whatever you want.   The code review process is too long.

Reply:
Tidy is ridiculous.

Reply:
> Overloading get_current_device in return types does not compile.

Not overloading on return types, overloading on `this` pointer qualification. That is doing:

```cpp
hip_device& get_current_device()
{
    assert(current_device != nullptr);
    return *current_device;
}
const hip_device& get_current_device() const
{
    assert(current_device != nullptr);
    return *current_device;
}
```

That will compile.

Reply:
const  hip_device& get_current_device() const
    {
        assert(current_device != nullptr);
        return *current_device;
    }

    hip_device& get_current_device()
    {
        assert(current_device != nullptr);
        return *current_device;
    }

    void finish() const
    {
        get_current_device().stream_sync();
        gpu_sync();
    }


/root/AMDMIGraphX/src/targets/gpu/include/migraphx/gpu/context.hpp:180:9: error: 'this' argument to member function 'stream_sync' has type 'const migraphx::version_1::gpu::hip_device', but function is not marked const

Reply:
I added the overload which leads to another error as shown above.

Reply:
You need to make `stream_sync` const.

Reply:
After making "stream_sync" const.  I got an error as shown below.   Note that you can't make "get" const since "get" it does change the data fields.

Besides, if we can make "stream_sync" const, why can't we make "record_event", "wait_event" const?

This is an example that Tidy has limitations.   It can only give static analysis.   It is ridiculous we have to change the code just to satisfy Tidy's limitation.



/root/AMDMIGraphX/src/targets/gpu/include/migraphx/gpu/context.hpp:135:42: error: 'this' argument to member function 'get' has type 'const __gnu_cxx::__alloc_traits<std::allocator<migraphx::version_1::gpu::hip_device::stream> >::value_type' (aka 'const migraphx::version_1::gpu::hip_device::stream'), but function is not marked const
                    hipStreamSynchronize(streams.at(i).get());
                                         ^~~~~~~~~~~~~
/root/AMDMIGraphX/src/targets/gpu/include/migraphx/gpu/context.hpp:43:21: note: 'get' declared here
        hipStream_t get()

Reply:
> Besides, if we can make "stream_sync" const, why can't we make "record_event", "wait_event" const?

We can also. The problem is with making `set_stream` const. That method shouldn't be const.

Why do we even need `stream_sync`? `hipDeviceSynchronize` will sync every stream on the device. All the tests pass if I remove that function.

Reply:
strream_sync synchronize all streams.  hipDeviceSynchronize doesn't do that.  If you run unit tests, they do not test runtime behavior.    I remember if we do not synchronize stream, we will run into runtime error when hipDeviceSynchronize is called.

Reply:
All these methods should not be const.  They were made const just to satisfy tidy.
Existing code already passes all tests.  Please merge.    It makes no sense to be stuck in this.

Reply:
Where are rnn tests?  You can file an item and assign it to me to investigate what the problem is after the check-in.   It makes no sense to refuse the merge until every corner is touched.

Reply:
> strream_sync synchronize all streams. hipDeviceSynchronize doesn't do that.

Yes it does, see [here](https://github.com/ROCm-Developer-Tools/HIP/blob/0dd26b4f63fdca3436ab686f62028918be28ceef/src/hip_hcc.cpp#L1084). Although, it does ignore the `hipStreamNonBlocking` so it might be a problem if we were using that flag which we are not.

Lets just remove the call to `stream_sync` for now in `finish`, and we can fix it later.

> All these methods should not be const. They were made const just to satisfy tidy.

Yes, because we dont call anything non-const, it suggest that the method be const. We can fix that by overloading the `get_current_device` with a const and non-const overload, so the const-ness of the `hip_stream` methods will propagate to `context`.

> Existing code already passes all tests.

They dont pass if I do `MIGRAPHX_DISABLE_NULL_STREAM=1 make check`:

```
99% tests passed, 1 tests failed out of 150

Total Test time (real) =  13.60 sec

The following tests FAILED:
	 28 - test_gpu_miopen (Failed)
```

> Where are rnn tests?

In test/miopen.cpp file.

Reply:
We are usiing hipStreamNonBlocking.  I did run into problems if I do not call stream_sync.  No, I can't remove it.

Reply:
I looked into miopen.cpp,  found that in the first failed test,  "gpu::pooling" has no finalize method.   stream execution depends on finalize to set up stream for MIOpen.